### PR TITLE
Add LegendItem

### LINKED ISSUE #5202: Figure legend not merging glyphs on the same data
So i have a legend created as follows:

```
    fig.line(x="Week of", y="% Under 14", source=source, line_width=2, legend="% Under 14 Days", line_color="blue")
    fig.circle(x="Week of", y="% Under 14", source=source, fill_color="white", size=8, legend="% Under 14 Days", line_color="blue")  
    fig.line(x="Week of", y="% Under 21", source=source, line_width=2, line_color="red", legend="% Under 21 Days")
    fig.circle(x="Week of", y="% Under 21", source=source, fill_color="white", size=8, legend="% Under 21 Days", line_color="red")  
```

Like in the example at http://bokeh.pydata.org/en/latest/docs/gallery/legend.html I would expect my legend to look something like:

image

However, my legend ends up looking like this:

image

Seems like a bug with the legend and the circles not being joined with the lines

### PR DESCRIPTION
**Pull request by birdsarah**
All pull requests must have an associated issue in the issue tracker. If there

- [x] issues: fixes #5202 #4526
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

Still TODO:
- [x] Add validation on legend items #5228
- [x] Remaining deprecations as per #4526

### CODE REVIEW

**Comment by birdsarah** [comment]
Have checked with JS API (only works after merging #5219). Here's the output

screen shot 2016-09-25 at 1 43 06 pm

File: bokeh/models/annotations.py [hunk 1]
```diff
@@ -35,19 +44,69 @@ class Annotation(Renderer):
...
+
+    @validation.error(NON_MATCHING_DATA_SOURCES_ON_LEGEND_ITEM_RENDERERS)
+    def _check_data_sources_on_renderers(self):
+        if self.label and self.label.get('field'):
+            if len(self.renderers) < 1:
+                return str(self)
+            source = self.renderers[0].data_source
+            for r in self.renderers:
+                if r.data_source != source:
+                    return str(self)
```
Comment (bryevdv) [inline bokeh/models/annotations.py]: I think this is equivalent and a bit simpler:

```
def _check_data_sources_on_renderers(self):
    if self.label and 'field' in self.label: 
        if len({ r.data_source for r in self.renderers }) != 1: 
            return str(self)
```

File: bokeh/plotting/helpers.py [hunk 2]
```diff
@@ -83,65 +88,66 @@ def _pop_colors_and_alpha(glyphclass, kwargs, prefix="", default_alpha=1.0):
...
         # let any non-dataspecs do their own validation (e.g., line_dash properties)
-        if var not in dataspecs: continue
-
+        if var not in dataspecs:
+            continue
         # strings sequences are handled by the dataspec as-is
-        if isinstance(val, string_types): continue
-
+        if isinstance(val, string_types):
+            continue
```
Comment (bryevdv) [inline bokeh/plotting/helpers.py]: I really prefer the old version, I find the whitespace between each pair of lines much more readable

File: examples/models/population_server.py [hunk 3]
```diff
@@ -48,20 +49,24 @@ def pyramid():
     female_quad = Quad(left=0, right="female", bottom="groups", top="shifted", fill_color="#CFF09E")
     female_quad_glyph = plot.add_glyph(source_pyramid, female_quad)
 
-    plot.add_layout(Legend(legends=[("Male", [male_quad_glyph]), ("Female", [female_quad_glyph])]))
+    plot.add_layout(Legend(items=[
+        LegendItem(label="Male", renderers=[male_quad_glyph]),
+        LegendItem(label="Female", renderers=[female_quad_glyph])
+    ]))
```
Comment (bryevdv) [inline examples/models/population_server.py]: What about adding one of those converters to the `items` property so that
```
plot.add_layout(Legend(items=[("Male", [male_quad_glyph]), ("Female", [female_quad_glyph])]))
```
works and converts to a list of `LegendItems` automatically?

File: sphinx/source/docs/user_guide/annotations.rst [hunk 4]
```diff
@@ -78,6 +78,12 @@ if needed by passing a legend that is the column of the column data source.
 .. bokeh-plot:: source/docs/user_guide/source_examples/plotting_legends_by_source.py
     :source-position: above
 
+If you do not want this auto behavior. You can:
+ * use the ``field()`` or ``value()`` functions from :ref:`bokeh.core.properties` to
+   be explicit about your intentions
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/annotations.rst]: a concrete example always helps users

File: sphinx/source/docs/user_guide/annotations.rst [hunk 5]
```diff
@@ -78,6 +78,12 @@ if needed by passing a legend that is the column of the column data source.
...
+ * not specify a legend argument and manually build a Legend yourself. There are examples of this
+   in the examples directory.
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/annotations.rst]: worth a link to one with `:bokeh-tree:`

File: bokehjs/src/coffee/models/annotations/legend_item.coffee [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  initialize: (attrs, options) ->
+    super(attrs, options)
+    # Validate data_sources match
+    data_source_validation = @_check_data_sources_on_renderers()
+    if not data_source_validation
+      logger.error("Non matching data sources on legend item renderers")
+    # Validate label in data_source
+    field_validation = @_check_field_label_on_data_source()
+    if not field_validation
+      logger.error("Bad column name on label: #{@label}")
```
Comment (bryevdv) [inline bokehjs/src/coffee/models/annotations/legend_item.coffee]: do these raise exceptions?

File: bokeh/plotting/helpers.py [hunk 7]
```diff
@@ -491,7 +525,8 @@ def func(self, *args, **kwargs):
     kwlines = []
     kws = glyphclass.properties() - set(glyphclass._args)
     for kw in sorted(kws):
-        if kw == "session": continue # TODO (bev) improve or remove
+        if kw == "session":
+            continue  # TODO (bev) improve or remove
```
Comment (bryevdv) [inline bokeh/plotting/helpers.py]: I think these two lines are complete cruft

File: bokehjs/src/coffee/models/annotations/legend_item.coffee (hunk 6 above)
Comment (birdsarah) [inline bokehjs/src/coffee/models/annotations/legend_item.coffee]: No, they do a console.error - so you get the same kind of warning you will see python side - aggressive but doesn't stop execution.

File: bokeh/models/annotations.py (hunk 1 above)
Comment (birdsarah) [inline bokeh/models/annotations.py]: ok

Comment (birdsarah) [inline bokeh/models/annotations.py]: thanks!

File: bokeh/plotting/helpers.py (hunk 2 above)
Comment (birdsarah) [inline bokeh/plotting/helpers.py]: lets give pep8 some love :D

File: bokeh/plotting/helpers.py (hunk 7 above)
Comment (birdsarah) [inline bokeh/plotting/helpers.py]: ok, i was just pep8-ing, this has nothing to do with this code. i'm not going to remove some code in this PR. Happy to do it in a follow-on though.

File: examples/models/population_server.py (hunk 3 above)
Comment (birdsarah) [inline examples/models/population_server.py]: I think that'd be great - can I follow on with another PR - so we can maybe get this out soon.

File: sphinx/source/docs/user_guide/annotations.rst (hunk 4 above)
Comment (birdsarah) [inline sphinx/source/docs/user_guide/annotations.rst]: good point

File: sphinx/source/docs/user_guide/annotations.rst (hunk 5 above)
Comment (birdsarah) [inline sphinx/source/docs/user_guide/annotations.rst]: ok

File: examples/models/population_server.py (hunk 3 above)
Comment (birdsarah) [inline examples/models/population_server.py]: I think there's a few variations on this which could make it quicker to spell / might be people's guess:

List on glyph_renderer optional:  
```python
items = [("Male", male_quad_glyph), ("Female", female_quad_glyph)]
```

Dictionary (a few ways)

```python
items = {"Male": male_quad_glyph, "Female", female_quad_glyph}
```

```python
items = {"Male": [male_quad_glyph], "Female", [female_quad_glyph]}
```

```python
items = [{"Male": [male_quad_glyph]}, {"Female", [female_quad_glyph]}]
```

Comment (bryevdv) [inline examples/models/population_server.py]: Problem with dict is that it is unordered.  And I don't like the list of dicts because it seems much more complicated and easy to get wrong for no value. Also list of tuples is whats already there it's a smaller step for users to migrate.

File: bokeh/models/annotations.py [hunk 8]
```diff
@@ -35,19 +44,64 @@ class Annotation(Renderer):
...
+            if len({r.data_source for r in self.renderers}) != 1:
+                return str(self)
+
+    @validation.error(BAD_COLUMN_NAME)
+    def _check_field_label_on_data_source(self):
+        if self.label and 'field' in self.label:
+            if len(self.renderers) < 1:
+                return str(self)
+            source = self.renderers[0].data_source
+            if self.label.get('field') not in source.column_names:
```
Comment (mattpap) [inline bokeh/models/annotations.py]: It's not a problem with this PR, but whenever I see `prop.get('field')`, knowing that `prop` is `StringSpec` or similar, my head boils. We need to get rid of all those dictionaries that attempt to act like structs.

File: bokeh/models/annotations.py [hunk 9]
```diff
@@ -103,33 +157,72 @@ class Legend(Annotation):
...
-    .. note::
-        The ``legends`` attribute may also be set from a dict or OrderedDict.
-        If a dict is used, the order of the legend entries is unspecified.
+    items = List(Instance(LegendItem), help="""
+    A list of legend items to be rendered in the legend.
+    """)
+
+    @property
+    def legends(self):
+        warnings.warn(DEP_MSG_0_12_3 % ('legends', 'Legend.items'))
```
Comment (mattpap) [inline bokeh/models/annotations.py]: You have to use `BokehDeprecationWarning`. Otherwise most users won't see this. Apply in other instances as well.

File: bokeh/plotting/helpers.py (hunk 2 above)
Comment (mattpap) [inline bokeh/plotting/helpers.py]: > lets give pep8 some love

This may bring some hating into this PR ;)

File: bokehjs/src/coffee/models/annotations/legend_item.coffee [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    # Always return a list of the labels
+    if @label? and @label.value?
+      return [@label.value]
+    field = @get_field_from_label_prop()
+    if field?
+      if @renderers[0] and @renderers[0].data_source?
+        source = @renderers[0].data_source
+      else
+        return ["No source found"]
+      if source.get_column?
```
Comment (mattpap) [inline bokehjs/src/coffee/models/annotations/legend_item.coffee]: Do `instanceof ColumnDataSource` instead (I presume this is the intent). If you leave this, next refactoring or "refactoring" (the term `refactoring` is misused all the time) around sources will be fun.

Comment (mattpap) [inline bokehjs/src/coffee/models/annotations/legend_item.coffee]: This pretty much applies to all uses of `?` operator. If you can avoid it and take advantage of the object model, do it. It states the intent better and is easier on future changes. Also, if you can differentiate between `null` and `undefined`, do it as well (so use e.g. `!= null` instead of `?`). A lot of people in JS community blend those two.

File: bokehjs/src/coffee/models/annotations/legend_item.coffee [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        data = source.get_column(field)
+        if data
```
Comment (mattpap) [inline bokehjs/src/coffee/models/annotations/legend_item.coffee]: On the contrary, here, `?` operator would be in order.

File: bokeh/plotting/helpers.py (hunk 7 above)
Comment (mattpap) [inline bokeh/plotting/helpers.py]: I strongly suggest limiting unrelated changes, because you may cause your self a lot of work in merge conflict resolution, and may bias reviewers to not focus on the subject of a PR (I know we shouldn't, but we are just people, usually with strong opinions).

**Comment by birdsarah** [comment]
Can u provide me an example of this. I copied this pattern from Plot. I had assumed, probably wrongly, the __deprecated_attributes__ property did the BokehDeprecationWarning

Sarah Bird
sbird@continuum.io

**Comment by mattpap** [comment]
@birdsarah, `__deprecated_attributes__` aren't about deprecation warnings, but internal bookkeeping in `HasProps`. You have to warn on your own, which you do, just instead of `warnings.warn(msg)` you issue `warnings.warn(msg, BokehDeprecationWarning)` (perhaps with `stacklevel=n` where `n` has to be figured out on case by case basis).

**Comment by birdsarah** [comment]
Thanks so much @mattpap, and thanks for the quick turnaround on a review.

**Comment by mattpap** [comment]
@birdsarah, after PR #5237 there will be only one way do deprecate things.

**Comment by bryevdv** [comment]
@birdsarah I think this should go into `0.12.3` since we have waited, thoughts?

**Comment by bryevdv** [comment]
@birdsarah I know you've got a lot going on with other work, I will try to merge master and address @mattpap couple of comments, I'd like to get this in for `0.12.3` later this week

File: bokeh/models/annotations.py (hunk 8 above)
Comment (bryevdv) [inline bokeh/models/annotations.py]: @mattpap more precisely, they are acting like unions. If you have a better proposal to support something like a union it is welcome.

**Comment by mattpap** [comment]
Merge conflicts are trivial.

**Comment by mattpap** [comment]
btw. If you want to quickly edit all conflicting files, then this may help (replace `vim -p` with your favorite editor):
```
[alias]
    resolve-merge-conflict = !sh -c 'vim -p $(git status -s | grep '^UU' | cut -c4-)
```
Add it to your `.gitconfig` and then you can issue `git resolve-merge-conflicts`.

**Comment by bryevdv** [comment]
@mattpap that should be added to the dev guide, that's great. Yes the conflicts are trivial. I am fixing those, making the few changes you suggested, then I was going to add support for auto-converting the "list of tuples" in the property for `.items`

**Comment by bryevdv** [comment]
@mattpap @birdsarah there are possibly some minor issues but I think this is ready for now with some followup issues:

* I had to change the `taylor_server.py` example, creating a new `LegendItem` with existing renderers inside was not deserializing correctly.  But I think that is possibly a weird case in general with server comms. In any case it's work-aroundable as seen. 

* I'm not sure about the JS side `LegendItem` logic. It's also work-aroundable. I had to add `value` when I didn't think I'd need to

File: bokehjs/test/models/annotations/legend_item.coffee [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+  describe "get_field_from_label_prop", ->
+
+    it "should return undefined if label property is null", ->
+      legend_item = new LegendItem({'label': null})
+      field = legend_item.get_field_from_label_prop()
+      expect(field).to.be.undefined
+
+    it "should return undefined if label property is value", ->
+      legend_item = new LegendItem({'label': {'value': 'milk'}})
```
Comment (mattpap) [inline bokehjs/test/models/annotations/legend_item.coffee]: This code was copied, so no blame here, but I would like to point out that it's preferred it use name literals instead of strings when referring to members of an interface, i.e. `new LegendItem({label: {value: 'milk'}})`. This states the intent better, because, after all, `label` and `value` are members of some interfaces and we aren't building an arbitrary dictionary here. Currently implicit, but when we will use TypeScript, it will be defined somewhere. Actually, it is already under `api/typings`. This is similar to what we do in Python when constructing keyword arguments with `dict(name=value)`, instead of `{"name": value}`.

Comment (bryevdv) [inline bokehjs/test/models/annotations/legend_item.coffee]: @mattpap that makes sense, I can push updates to those as well. Do you have any other comments on this PR?

File: bokeh/models/annotations.py [hunk 13]
```diff
@@ -35,19 +44,64 @@ class Annotation(Renderer):
...
+    def _check_field_label_on_data_source(self):
+        if self.label and 'field' in self.label:
+            if len(self.renderers) < 1:
+                return str(self)
+            source = self.renderers[0].data_source
+            if self.label.get('field') not in source.column_names:
+                return str(self)
+
+
+DEP_MSG_0_12_3 = """
```
Comment (mattpap) [inline bokeh/models/annotations.py]: Deprecation warnings have to be changed the new style.

**Comment by mattpap** [comment]
> It's also work-aroundable. I had to add value when I didn't think I'd need to.

Is this an issue with js examples or test?

**Comment by bryevdv** [comment]
> Is this an issue with js examples or test?

Actually I think it just needs to be expicit for `models` examples. The nice convenience checking for field of value is part of the `plotting` API

**Comment by mattpap** [comment]
LGTM. Any inconveniences will emerge sooner than later after merge.

**Comment by birdsarah** [comment]
@bryevdv @mattpap, thanks for taking this PR the last mile - have looked at all the commits that @bryevdv made and they look good - I'd forgotten about "accepts" so that was much easier than I thought it was going to be.

@mattpap, really appreciating the explanations on PR comments I'm learning a lot and hopefully you won't have to repeat yourself in the future.

**Comment by bryevdv** [comment]
OK I will merge this when green, I'll make issues for the other small questions I ran into.

**Comment by bryevdv** [comment]
I can't get the py27 unit tests to pass, the port in use error seems persistent. I wish I knew what changed on travis or whatever else to start this problem.

**Comment by mattpap** [comment]
@bryevdv, is this `5006` port in use error? I complained about this on internal mailing list. This error comes and goes, so you can restart a dozen times with no result and then after an hour or two tests will start to work again.

**Comment by bryevdv** [comment]
Yes, it is the same you mentioned. What is odd is that it is always the same test (`test_log_status`)

**Comment by bryevdv** [comment]
trying specifying a different port in `ManagedServerLoop` (which BTW is used all over the tests, its very strange that only this one test has issue)

**Comment by bryevdv** [comment]
that just kicks the can, now a different test is failing the same way. It's always just one.

**Comment by bryevdv** [comment]
It's also only python 2.7

**Comment by bryevdv** [comment]
OK, `--rerun` "worked", will merge when green
